packages/andive/src/components/direction/destination.tsx

```diff
@@ -70,31 +69,86 @@ const AsideLabel = styled(({ offsetY, ...props }) => <Body1 {...props} />)`
   text-align: right;
 `
 
+type DestinationContextType = {
+  ref: React.Ref<HTMLElement | null>
+} | null
+
+const DestinationContext = React.createContext<DestinationContextType>(null)
+
+type PointProps = {
+  children: React.ReactElement
+}
+
+function Point(props: PointProps) {
+  const contextValue = React.useContext(DestinationContext)
+  return React.cloneElement(props.children, contextValue ? {ref: contextValue.ref} : undefined)
+}
+
 interface DestinationProps<PointRefElementType> {
-  className?: string,
-  label?: string,
-  children: React.ReactNode | ((ref: React.Ref<HTMLDivElement>, pointRef: React.Ref<PointRefElementType>) => React.ReactNode)
+  className?: string
+  label?: string
+  children: React.ReactNode
 }
 export function Destination<PointRefElementType>({label, children, ...props}: DestinationProps<PointRefElementType>) {
-  const ref = useRef(null)
-  const size = useElementRect(ref)
-  const pointRef = useRef(null)
-  const pointSize = useElementRect(pointRef)
+  const [rect, setRect] = React.useState<DOMRect | ClientRect>()
+  const [pointRect, setPointRect] = React.useState<DOMRect | ClientRect>()
+  const ref = React.useRef<HTMLDivElement>(null)
+  const pointRef = React.useRef<HTMLElement>(null)
+
+  React.useLayoutEffect(() => {
```

Comment:
From https://kentcdodds.com/blog/useeffect-vs-uselayouteffect

> This runs synchronously immediately after React has performed all DOM mutations. This can be useful if you need to make DOM measurements (like getting the scroll position or other styles for an element) and then make DOM mutations or trigger a synchronous re-render by updating state.

> As far as scheduling, this works the same way as componentDidMount and componentDidUpdate. Your code runs immediately after the DOM has been updated, but before the browser has had a chance to "paint" those changes (the user doesn't actually see the updates until after the browser has repainted).

Looks like it's the way to go because you call a setRect/setPointRect with a node measurement (getBoundingClientRect).

I just fail to understand how a setRect() can be applied before the browser re-paint. But maybe React can postpone the repaint by enqueuing a microtask 🤔

Reply:
https://raw.githubusercontent.com/donavon/hook-flow/master/hook-flow.png